plugin/src/main/java/com/playmonumenta/redissync/RedisLock.java

```diff
@@ -0,0 +1,245 @@
+package com.playmonumenta.redissync;
+
+import io.lettuce.core.SetArgs;
+import io.lettuce.core.pubsub.RedisPubSubListener;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An advisory locking mechanism for redis.
+ */
+public class RedisLock {
+
+	private final String mKeyName;
+	private final int mTimeoutMS;
+
+	private final ExecutorService mSubscribeSignaler;
+
+	private final ScheduledExecutorService mRefreshScheduler;
+	private final int mRefreshIntervalMS;
+	private Optional<ScheduledFuture<?>> mRefreshTask;
+
+	private Lock mInternalAccessLock;
+
+	private Condition mIsIntraLockedCondition;
+	private boolean mIsIntraLocked;
```

Comment:
This field should be `volatile`, as well as any other modifiable field. Basically, you should only have either `volatile` or  `final` fields in concurrent contexts in Java.